File: solutions/c/rational-numbers/1/rational_numbers.c
// Dedicated to Shree DR.MDD — eternal guidance in code

#include "rational_numbers.h"
#include <math.h>
#include <stdlib.h>

rational_t add(rational_t a, rational_t b){
    rational_t res = {
        a.numerator * b.denominator + b.numerator * a.denominator,
        a.denominator * b.denominator
    };
    return reduce(res);
}

rational_t subtract(rational_t a, rational_t b) {
    rational_t res = {
        a.numerator * b.denominator - b.numerator * a.denominator,
        a.denominator * b.denominator
    };
    return reduce(res);
}

rational_t multiply(rational_t a, rational_t b) {
    rational_t res = {
        a.numerator * b.numerator,
        a.denominator * b.denominator
    };
    return reduce(res);
}

rational_t divide(rational_t a, rational_t b) {
    rational_t res = {
        a.numerator * b.denominator,
        a.denominator * b.numerator
    };
    return reduce(res);
}

rational_t absolute(rational_t a) {
    rational_t res = {
        abs(a.numerator),
        abs(a.denominator)
    };
    return reduce(res);
}

rational_t exp_rational(rational_t a, int16_t b) {
    if (b < 0) {
        rational_t inv = { a.denominator, a.numerator };
        return exp_rational(inv, -b);
    }
    rational_t res = { 1, 1 };
    for (int i = 0; i < b; ++i) {
        res.numerator *= a.numerator;
        res.denominator *= a.denominator;
    }
    return reduce(res);
}

float exp_real(float a, rational_t b) {
    return pow(pow(a, b.numerator), 1.0f / (float)b.denominator);
}

rational_t reduce(rational_t a) {
    int p = abs(a.numerator);
    int q = abs(a.denominator);
    if (!p) {
        a.denominator = 1;
        return a;
    }
    int temp;
    while (q) { temp = q; q = p % q; p = temp; }
    a.numerator /= p;
    a.denominator /= p;
    if (a.denominator < 0) {
        a.numerator = -a.numerator;
        a.denominator = -a.denominator;
    }
    return a;
}

File: solutions/go/rational-numbers/1/rational_numbers.go
package rationalnumbers

import "math"

type Rational struct {
	numerator, denominator int
}

func abs(n int) int {
	if n < 0 {
		return -n
	}
	return n
}

func gcd(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

func pow(base, exp int) int {
	return int(math.Pow(float64(base), float64(exp)))
}

func (r Rational) Reduce() Rational {
	gcd := gcd(abs(r.numerator), abs(r.denominator))
	numerator := r.numerator / gcd
	denominator := r.denominator / gcd

	if denominator < 0 {
		numerator = -numerator
		denominator = -denominator
	}

	return Rational{numerator, denominator}
}

func (r Rational) Add(s Rational) Rational {
	numerator := r.numerator*s.denominator + s.numerator*r.denominator
	denominator := r.denominator * s.denominator

	return Rational{numerator, denominator}.Reduce()
}

func (r Rational) Sub(s Rational) Rational {
	return r.Add(Rational{-s.numerator, s.denominator})
}

func (r Rational) Mul(s Rational) Rational {
	return Rational{r.numerator * s.numerator, r.denominator * s.denominator}.Reduce()
}

func (r Rational) Div(s Rational) Rational {
	return r.Mul(Rational{s.denominator, s.numerator})
}

func (r Rational) Abs() Rational {
	return Rational{abs(r.numerator), abs(r.denominator)}.Reduce()
}

// Compute r ^ power, a rational raised to an int exponent.
func (r Rational) Exprational(power int) Rational {
	if power >= 0 {
		return Rational{pow(r.numerator, power), pow(r.denominator, power)}.Reduce()
	} else {
		return Rational{pow(r.denominator, -power), pow(r.numerator, -power)}.Reduce()
	}
}

// Compute base ^ r, an int raised to a rational.
func (r Rational) Expreal(base int) float64 {
	return math.Pow(float64(base), float64(r.numerator)/float64(r.denominator))
}